crates/tabby/src/services/code.rs

```diff
@@ -151,27 +152,40 @@ fn closest_match<'a>(
     search_term: &'a str,
     search_input: impl IntoIterator<Item = &'a RepositoryConfig>,
 ) -> Option<String> {
-    let search_term = RepositoryConfig::canonicalize_url(search_term);
-
-    let mut nucleo = nucleo::Matcher::new(nucleo::Config::DEFAULT.match_paths());
+    let git_search = GitUrl::parse(search_term).ok()?;
     search_input
         .into_iter()
-        .filter_map(|entry| {
-            let url = entry.canonical_git_url();
-            Some((
-                url.clone(),
-                // Matching using the input URL as the haystack instead of the needle yielded better scoring
-                // Example:
-                // haystack = "https://github.com/boxbeam/untwine" needle = "https://abc@github.com/boxbeam/untwine.git" => No match
-                // haystack = "https://abc@github.com/boxbeam/untwine.git" needle = "https://github.com/boxbeam/untwine" => Match, score 842
-                nucleo.fuzzy_match(
-                    Utf32String::from(&*search_term).slice(..),
-                    Utf32String::from(&*url).slice(..),
-                )?,
-            ))
+        .filter_map(|elem| {
+            let Ok(git_url) = GitUrl::parse(&elem.git_url) else {
+                return None;
+            };
+            let mut score = 0;
+            // Name is scored the highest - more than 3 letters different automatically disqualifies
+            score += 15 * damerau_levenshtein(&git_search.name, &git_url.name);
+
+            // Host is scored the second highest, but it is okay for the host to be different if
+            // the name is very similar or a perfect match
+            if let Some((host1, host2)) = git_search.host.as_ref().zip(git_url.host) {
+                score += 4 * damerau_levenshtein(host1, &host2);
+            } else {
+                score += 30;
+            }
+
+            // Owner is scored the lowest, to allow forks to be matched as long as the rest matches
+            if let Some((org1, org2)) = git_search.owner.as_ref().zip(git_url.owner) {
+                score += 3 * damerau_levenshtein(org1, &org2);
+            } else {
+                score += 15;
+            }
```

Comment:
as discussed, don't use scoring in first step. Find suffix matching (until first deliminator) to find candidates, and only do scoring if there're multiple candidates